### ISSUE TITLE
MLX: adapter_config.json saves unresolved `target_modules: auto` — adapter loads as a silent no-op in mlx_lm

### ISSUE THREAD

**Issue by armanbot-jpg** [MakazhanAlpamys/Soup#392]
## What happened

Trained SFT LoRA with `backend: mlx` (git main, post-#362) on an M-series Mac:

```yaml
base: mlx-community/Qwen2.5-0.5B-Instruct-4bit
task: sft
backend: mlx
training:
  epochs: 4
  lr: 1e-4
  batch_size: 2
  lora: {r: 16, alpha: 32}
output: ./output
```

Training works (loss 3.22 → 0.78, adapter saved). But generation with the adapter is **bit-identical to the base model**, with no warning:

```bash
python -m mlx_lm generate --model mlx-community/Qwen2.5-0.5B-Instruct-4bit \
  --adapter-path output --prompt "..."   # == base output
```

## Root cause

`_apply_lora()` resolves `target_modules: auto` into a local variable (`mlx_sft.py:109-110`):

```python
target_keys = lora_cfg.target_modules
if not target_keys or target_keys in (["auto"], "auto"):
    target_keys = ["self_attn.q_proj", "self_attn.v_proj"]
```

…but the `adapter_config.json` writer (`mlx_sft.py:236-239`) serializes the **raw config value**, not the resolved list:

```python
"keys": list(
    lora_cfg.target_modules
    if isinstance(lora_cfg.target_modules, list)
    else [lora_cfg.target_modules]
),
```

With the default `target_modules: auto`, the shipped config contains `"keys": ["auto"]`. On load, `mlx_lm.tuner.utils.load_adapters` calls `linear_to_lora_layers(..., {"keys": ["auto"]})`, which matches no module, then `model.load_weights(strict=False)` silently drops all 96 LoRA tensors. The comment above the writer says the file exists precisely so the dir is "directly loadable with mlx_lm.load(..., adapter_path=output)" — that promise is what breaks.

This is the same failure class as the v0.72.0 `.inner.` inert-adapter bug and #362's full-fine-tune adapter: training looks healthy, artifact is silently dead.

## Verification

Editing the saved `adapter_config.json` by hand:

```json
"keys": ["self_attn.q_proj", "self_attn.v_proj"]
```

…makes the very same `adapters.safetensors` produce the fine-tuned behavior. One-line fix: persist the resolved `target_keys` from `_apply_lora` (e.g. stash `self._resolved_target_keys`) instead of `lora_cfg.target_modules`.

## Two smaller things hit on the way (can split into separate issues)

1. **PyPI v0.73.0 predates #362** — `pip install soup-cli` still dispatches `backend: mlx` through the transformers path, which crashes loading an MLX 4-bit checkpoint (`Embedding: [151936, 112] vs [151936, 896]`). If a patch release is planned, this bug would be worth riding along.
2. **`pip install "soup-cli[mlx]"` alone cannot train** — the trainer requires `datasets`, then TRL, before reaching the MLX branch, so docs' "[mlx] is enough" install path fails; full `[train]` (torch included) is required in practice.

Environment: macOS arm64 (M-series, 16 GB), Python 3.12.12, mlx 0.30.x, mlx-lm 0.31.x, soup-cli git main @ 9f57b7c.

**Comment by MakazhanAlpamys** [comment]
Confirmed and fixed in [`50e1865`](https://github.com/MakazhanAlpamys/Soup/commit/50e1865). Verified every claim against the source rather than taking the report on trust; all of them hold.

`mlx_sft.py` resolved into a local at the `_apply_lora` site and serialised `lora_cfg.target_modules` at the writer, exactly as you traced. And `target_modules` defaults to **`"auto"`** in the schema, so this was **every MLX run that did not name its modules by hand** — not an edge case.

**The control is what made this actionable.** Hand-editing `keys` and getting tuned behaviour from the *same* `adapters.safetensors` separates "the weights are wrong" from "the config is wrong" in one step, and it is the difference between a report I have to reproduce from scratch and one I can act on immediately. Most reports of "the adapter does nothing" do not do that.

You also placed it correctly: this is the third time in this project a healthy loss curve has shipped a dead artifact — v0.72.0's `.inner.` adapter keys, #362's full-fine-tune adapter, and now this. The loss curve structurally cannot see any of them; only the artifact can.

**The fix is one shared resolver, not a second copy of the resolution.** `resolve_mlx_target_keys()` is now the single answer to "which modules does this run train?", called by both `_apply_lora` and the config writer, because two copies of that question is how they drifted. Ten tests, red before and green after, including two controls: the resolved value must never contain `"auto"` for any input, and `rank` / `scale` / `dropout` must still come from the config — so a repair that hardcoded the whole block would fail instead of passing.

---

## Your two smaller items

**1. PyPI v0.73.0 predates #362 — you are right, and this is the more urgent half.** Anyone doing `pip install soup-cli` today still gets `backend: mlx` dispatched through the transformers path. So the *shipped* state is worse than the one you hit: you were on git main and got a silent no-op, a PyPI user gets the `Embedding: [151936, 112] vs [151936, 896]` crash. Both #362 and this fix are on `main` and unreleased. That is an argument for a patch release rather than waiting, and I am treating it as one.

**2. `pip install "soup-cli[mlx]"` cannot train — also correct, and it is a docs defect, not a packaging preference.** The trainer reaches `datasets` and TRL before the MLX branch, so `[mlx]` alone cannot get you to a training loop while the docs say it is enough. Filed separately rather than folded in here, since the fix is either a real dependency change or a documentation correction and the two want different discussions.

Thank you for this one — the report was better than the issue I would have written from the same symptom. If you want to take either follow-up yourself, say so and it is yours; [`CONTRIBUTING.md`](https://github.com/MakazhanAlpamys/Soup/blob/main/CONTRIBUTING.md) has the claiming rules, and there is no expectation either way.

**Comment by MakazhanAlpamys** [comment]
Main defect fixed in [`50e1865`](https://github.com/MakazhanAlpamys/Soup/commit/50e1865) — the resolved module list is what reaches `adapter_config.json` now, and one shared `resolve_mlx_target_keys()` is what both callers use.

Your second follow-up is **#394** (`[mlx]` alone cannot train — the MLX path imports `datasets` + TRL before reaching the MLX branch), filed with both candidate fixes and the measurement that should decide between them, credited to you. Yours if you want it.

The first (PyPI predating #362) is a release-timing matter rather than a code one, and it argues for a patch release rather than waiting — tracked as part of the next release, not as its own issue.

**Comment by MakazhanAlpamys** [comment]
Update, ~90 seconds after the fix landed: **it is already on PyPI.** `v0.73.1` was tagged on the commit that carries it, so `pip install --upgrade soup-cli` gets you both #362 (the dispatch fix, which removes the `Embedding: [151936, 112]` crash you named in your first follow-up) and this one.

That also closes your first item on its own — the PyPI/git-main gap you hit no longer exists. Worth re-running your original repro against `0.73.1` if you have the machine handy; the adapter should now change generation, and if it does not I would want to know immediately, because that would mean the resolved keys still are not reaching `mlx_lm`.

Only #394 (`[mlx]` alone cannot train) is still open from your report.
